File: rb/entity/get_location_by_postal_code_entity.rb
# ZippopotamusZipCode SDK GetLocationByPostalCode entity

require_relative '../utility/struct/voxgig_struct'
require_relative '../core/helpers'

class GetLocationByPostalCodeEntity
  def initialize(client, entopts = nil)
    entopts ||= {}
    if entopts["active"].nil?
      entopts["active"] = true
    elsif entopts["active"] == false
      # keep false
    else
      entopts["active"] = true
    end

    @_name = "get_location_by_postal_code"
    @_client = client
    @_utility = client.get_utility
    @_entopts = entopts
    @_data = {}
    @_match = {}

    @_entctx = @_utility.make_context.call({
      "entity" => self,
      "entopts" => entopts,
    }, client.get_root_ctx)

    @_utility.feature_hook.call(@_entctx, "PostConstructEntity")
  end

  def get_name
    @_name
  end

  def make
    opts = @_entopts.dup
    GetLocationByPostalCodeEntity.new(@_client, opts)
  end

  def data_set(args)
    if args
      @_data = ZippopotamusZipCodeHelpers.to_map(VoxgigStruct.clone(args)) || {}
      @_utility.feature_hook.call(@_entctx, "SetData")
    end
  end

  # @return [GetLocationByPostalCode, Hash] the current GetLocationByPostalCode data
  def data_get
    @_utility.feature_hook.call(@_entctx, "GetData")
    VoxgigStruct.clone(@_data)
  end

  def match_set(args)
    if args
      @_match = ZippopotamusZipCodeHelpers.to_map(VoxgigStruct.clone(args)) || {}
      @_utility.feature_hook.call(@_entctx, "SetMatch")
    end
  end

  # @return [Hash] the current match filter (any subset of GetLocationByPostalCode fields)
  def match_get
    @_utility.feature_hook.call(@_entctx, "GetMatch")
    VoxgigStruct.clone(@_match)
  end

  # Feature #4: run `action` through the full pipeline and return an Enumerator
  # over result items, so the `streaming` feature's incremental output is
  # reachable from a generated entity (a normal op call materialises the whole
  # result). `callopts` parameterises the call:
  #   - inbound (download): the Enumerator yields items/chunks (from the
  #     streaming feature when active, else the materialised items);
  #   - outbound (upload): an iterable `body` in callopts is attached to the
  #     request so the transport can stream the payload;
  #   - `ctrl` (pipeline control) and `signal` (cancellation) honoured.
  def stream(action, args = nil, callopts = nil)
    utility = @_utility
    callopts ||= {}
    signal = callopts["signal"]

    ctrl = callopts["ctrl"].is_a?(Hash) ? callopts["ctrl"].dup : {}
    ctrl["stream"] = callopts

    ctxmap = {
      "opname" => action,
      "ctrl" => ctrl,
      "match" => @_match,
      "data" => @_data,
    }
    args.each { |k, v| ctxmap[k] = v } if args.is_a?(Hash)

    ctx = utility.make_context.call(ctxmap, @_entctx)

    # Outbound: expose the caller's iterable payload so the request builder /
    # transport can stream it as the request body.
    body = callopts["body"]
    unless body.nil?
      ctx.reqdata["body$"] = body
      ctx.meta["stream_out"] = body
    end

    aborted = lambda do
      return false if signal.nil?
      return !!signal.call if signal.respond_to?(:call)
      return !!signal.aborted if signal.respond_to?(:aborted)
      false
    end

    Enumerator.new do |yielder|
      catch(:stream_stop) do
        utility.feature_hook.call(ctx, "PrePoint")
        point, err = utility.make_point.call(ctx)
        ctx.out["point"] = point
        throw :stream_stop if err

        utility.feature_hook.call(ctx, "PreSpec")
        spec, err = utility.make_spec.call(ctx)
        ctx.out["spec"] = spec
        throw :stream_stop if err

        utility.feature_hook.call(ctx, "PreRequest")
        resp, err = utility.make_request.call(ctx)
        ctx.out["request"] = resp
        throw :stream_stop if err

        utility.feature_hook.call(ctx, "PreResponse")
        resp2, err = utility.make_response.call(ctx)
        ctx.out["response"] = resp2
        throw :stream_stop if err

        utility.feature_hook.call(ctx, "PreResult")
        result, err = utility.make_result.call(ctx)
        ctx.out["result"] = result
        throw :stream_stop if err

        utility.feature_hook.call(ctx, "PreDone")

        result = ctx.result

        # Inbound: prefer the streaming feature's incremental Enumerator; else
        # fall back to the materialised items so stream always yields.
        stream_enum = result ? result.stream : nil
        if stream_enum
          stream_enum.each do |item|
            throw :stream_stop if aborted.call
            yielder << item
          end
        else
          data = utility.done.call(ctx)
          items = data.is_a?(Array) ? data : (data.nil? ? [] : [data])
          items.each do |item|
            throw :stream_stop if aborted.call
            yielder << item
          end
        end
      end
    end
  end

  

  
  # List GetLocationByPostalCode items matching the given filter.
  #
  # @param reqmatch [GetLocationByPostalCodeListMatch, Hash, nil] match filter (any subset of
  #   GetLocationByPostalCode fields); defaults to nil, treated as an empty match that lists all.
  # @param ctrl [Object, nil] optional per-call control
  # @return [Array<GetLocationByPostalCode>, Array] the matching GetLocationByPostalCode items; raises ZippopotamusZipCodeError on failure
  def list(reqmatch = nil, ctrl = nil)
    utility = @_utility
    ctx = utility.make_context.call({
      "opname" => "list",
      "ctrl" => ctrl,
      "match" => @_match,
      "data" => @_data,
      "reqmatch" => reqmatch,
    }, @_entctx)

    records = _run_op(ctx) do
      if ctx.result
        @_match = ctx.result.resmatch if ctx.result.resmatch
      end
    end

    # list yields the BARE Array of records — each an accessible Hash — so
    # callers can index item["id"] directly, matching py/lua/go. make_result
    # wraps each entry as an Entity instance for internal use; unwrap those
    # back to their bare record Hashes here (load/create/etc. are unaffected).
    if records.is_a?(Array)
      records = records.map do |item|
        item.respond_to?(:data_get) ? item.data_get : item
      end
    end

    records
  end



  

  

  

  private

  def _run_op(ctx, &post_done)
    utility = @_utility

    begin
      utility.feature_hook.call(ctx, "PrePoint")

      point, err = utility.make_point.call(ctx)
      ctx.out["point"] = point
      return utility.make_error.call(ctx, err) if err

      utility.feature_hook.call(ctx, "PreSpec")

      spec, err = utility.make_spec.call(ctx)
      ctx.out["spec"] = spec
      return utility.make_error.call(ctx, err) if err

      utility.feature_hook.call(ctx, "PreRequest")

      resp, err = utility.make_request.call(ctx)
      ctx.out["request"] = resp
      return utility.make_error.call(ctx, err) if err

      utility.feature_hook.call(ctx, "PreResponse")

      resp2, err = utility.make_response.call(ctx)
      ctx.out["response"] = resp2
      return utility.make_error.call(ctx, err) if err

      utility.feature_hook.call(ctx, "PreResult")

      result, err = utility.make_result.call(ctx)
      ctx.out["result"] = result
      return utility.make_error.call(ctx, err) if err

      utility.feature_hook.call(ctx, "PreDone")

      post_done.call

      utility.done.call(ctx)
    rescue StandardError => operr
      utility.feature_hook.call(ctx, "PreUnexpected")

      raise operr
    end
  end
end

File: ts/src/entity/GetLocationByPostalCodeEntity.ts

import { inspect } from 'node:util'

import { ZippopotamusZipCodeEntityBase } from '../ZippopotamusZipCodeEntityBase'

import type {
  ZippopotamusZipCodeSDK,
} from '../ZippopotamusZipCodeSDK'


import type {
  Operation,
  Context,
  Control,
} from '../types'

import type {
  GetLocationByPostalCode,
  GetLocationByPostalCodeListMatch,
} from '../ZippopotamusZipCodeTypes'

// TODO: needs Entity superclass
class GetLocationByPostalCodeEntity extends ZippopotamusZipCodeEntityBase<GetLocationByPostalCode> {

  constructor(client: ZippopotamusZipCodeSDK, entopts: any) {
    super(client, entopts)
    this.name = 'get_location_by_postal_code'
    this.name_ = 'get_location_by_postal_code'
    this.Name = 'GetLocationByPostalCode'
  }


  make(this: GetLocationByPostalCodeEntity) {
    return new GetLocationByPostalCodeEntity(this._client, this.entopts())
  }




  async list(this: any, reqmatch?: GetLocationByPostalCodeListMatch, ctrl?: Control): Promise<GetLocationByPostalCodeEntity[]> {

    const utility = this._utility

    const {
      makeContext,
      done,
      error,
      featureHook,
      makePoint,
      makeRequest,
      makeResponse,
      makeResult,
      makeSpec,
    } = utility

    let fres: Promise<any> | undefined = undefined

    let ctx: Context = makeContext({
      opname: 'list',
      ctrl,
      match: this._match,
      data: this._data,
      reqmatch
    }, this._entctx)

    try {

      fres = featureHook(ctx, 'PrePoint')
      if (fres instanceof Promise) { await fres }

      ctx.out.point = makePoint(ctx)
      if (ctx.out.point instanceof Error) {
        return error(ctx, ctx.out.point)
      }



      fres = featureHook(ctx, 'PreSpec')
      if (fres instanceof Promise) { await fres }

      ctx.out.spec = makeSpec(ctx)
      if (ctx.out.spec instanceof Error) {
        return error(ctx, ctx.out.spec)
      }



      fres = featureHook(ctx, 'PreRequest')
      if (fres instanceof Promise) { await fres }

      ctx.out.request = await makeRequest(ctx)
      if (ctx.out.request instanceof Error) {
        return error(ctx, ctx.out.request)
      }



      fres = featureHook(ctx, 'PreResponse')
      if (fres instanceof Promise) { await fres }

      ctx.out.response = await makeResponse(ctx)
      if (ctx.out.response instanceof Error) {
        return error(ctx, ctx.out.response)
      }



      fres = featureHook(ctx, 'PreResult')
      if (fres instanceof Promise) { await fres }

      ctx.out.result = await makeResult(ctx)
      if (ctx.out.result instanceof Error) {
        return error(ctx, ctx.out.result)
      }



      fres = featureHook(ctx, 'PreDone')
      if (fres instanceof Promise) { await fres }

      if (null != ctx.result) {
        if (null != ctx.result.resmatch) {
          this._match = ctx.result.resmatch
        }
      }

      return done(ctx)
    }
    catch (err: any) {

      fres = featureHook(ctx, 'PreUnexpected')
      if (fres instanceof Promise) { await fres }

      err = this._unexpected(ctx, err)

      if (err) {
        throw err
      }
      else {
        // Off-happy-path (throw disabled): typed as any so the method's
        // Promise<GetLocationByPostalCode[]> return stays clean under strict null checks.
        return undefined as any
      }
    }
  }





}


export {
  GetLocationByPostalCodeEntity
}
